File: comicboxd/web/src/js/models/book.ts
import * as grapql from 'js/graphql'

export interface Book {
    id: string
    alternate_series: string
    authors: string[]
    chapter: number
    community_rating: number
    created_at: string
    current_page: number
    date_released: string
    file: string
    genres: string[]
    last_page_read: number
    pages: Page[]
    rating: number
    read: Boolean
    reading_direction: string
    series: string
    story_arc: string
    summary: string
    title: string
    type: string
    updated_at: string
    volume: number
    web: string
}

export interface Page {
    file_number: number
    type: string
    url: string
}

export async function Find(id: string): Promise<Book> {
    let r = await grapql.Exec(`
    query getBook($id: ID!) {
        book(id: $id) {
            id
            alternate_series
            authors
            chapter
            community_rating
            created_at
            current_page
            date_released
            file
            genres
            last_page_read
            pages {
                file_number
                type
                url
            }
            rating
            read
            reading_direction
            series
            story_arc
            summary
            title
            type
            updated_at
            volume
            web
        }
    }      
    `, { id: id })
    return r.data.book
}

File: comicboxd/app/schema/book.go
package schema

import (
	"encoding/json"

	graphql "github.com/graph-gophers/graphql-go"

	"github.com/zwzn/comicbox/comicboxd/app/model"
)

type BookResolver struct {
	b model.BookUserBook
}

func (r *BookResolver) AlternateSeries() string {
	return r.b.AlternateSeries
}
func (r *BookResolver) Authors() []string {
	data := r.b.AuthorsJSON
	authors := []string{}
	json.Unmarshal(data, &authors)
	return authors
}

func (r *BookResolver) Chapter() *float64 {
	return r.b.Chapter
}

func (r *BookResolver) CommunityRating() *float64 {
	return r.b.CommunityRating
}

func (r *BookResolver) Cover() {

}

func (r *BookResolver) CreatedAt() graphql.Time {
	return graphql.Time{Time: r.b.CreatedAt}
}

func (r *BookResolver) CurrentPage() *int32 {
	if r.b.CurrentPage == nil {
		return nil
	}
	cp := int32(*r.b.CurrentPage)
	return &cp

}

func (r *BookResolver) DateReleased() *graphql.Time {
	date := r.b.DateReleased
	if date == nil {
		return nil
	}
	return &graphql.Time{Time: *date}
}
func (r *BookResolver) File() string {
	return r.b.File
}

func (r *BookResolver) Genres() []string {
	data := r.b.GenresJSON
	authors := []string{}
	json.Unmarshal(data, &authors)
	return authors
}

func (r *BookResolver) ID() graphql.ID {
	return graphql.ID(r.b.ID.String())
}
func (r *BookResolver) LastPageRead() {

}
func (r *BookResolver) Pages() {

}
func (r *BookResolver) Rating() {

}
func (r *BookResolver) Read() {

}
func (r *BookResolver) ReadingDirection() {

}
func (r *BookResolver) Series() {

}
func (r *BookResolver) SeriesQuery() {

}
func (r *BookResolver) StoryArc() {

}
func (r *BookResolver) Summary() {

}
func (r *BookResolver) Title() {

}
func (r *BookResolver) Type() {

}
func (r *BookResolver) UpdatedAt() {

}
func (r *BookResolver) Volume() {

}
func (r *BookResolver) Web() {

}
